File: 022-SimpleDijkstras/HandBookSolution.java
import java.util.*;

public class HandBookSolution {
    
    static final long INF = Long.MAX_VALUE;
    
    static class Entry implements Comparable {
        
        public int index;
        public long distance;
        
        public Entry(long d, int i) {
            distance = d;
            index = i;
        }

        public int compareTo(Object o) {
            
            Entry e = (Entry)o;
            
            if (distance != e.distance) {
                return (int)(distance - e.distance);
            }

            return index - e.index;
        }
    }

    static class Edge {
        
        public int index;
        public long length;
        
        public Edge(int i, long l) {
            index = i;
            length = l;
        }
    }

    static ArrayList<Edge> adj[];

    static void dijkstra(int N, int start, int goal) {
        
        final long dist[] = new long[N];
        Arrays.fill(dist, INF);
        
        dist[start] = 0;
        
        int prev[] = new int[N];
        
        Queue<Entry> frontier = new PriorityQueue<Entry>(dist.length);
        frontier.offer(new Entry(0, start));
        
        while (!frontier.isEmpty()) {
            
            Entry e = frontier.poll();
            int u = e.index;

            if (u == goal) {
                break;
            }

            if (dist[u] < e.distance) {
                continue;
            }

            for (Edge edge : adj[u]) {

                long uv = edge.length;
                int v = edge.index;

                if (dist[u] + uv < dist[v]) {

                    prev[v] = u;
                    dist[v] = dist[u] + uv;
                    frontier.offer(new Entry(dist[v], v));
                
                }
            }
        }

        System.out.println(dist[goal]);

    }

    public static void main(String[] args) {

        Scanner in = new Scanner(System.in);
        
        int N = in.nextInt();
        int E = in.nextInt();
        int S = in.nextInt();
        int T = in.nextInt();
        
        adj = (ArrayList<Edge>[])new ArrayList[N];
        
        for (int n = 0; n < N; n++) {
            adj[n] = new ArrayList<Edge>();
        }

        for (int e = 0; e < E; e++) {
            
            int a = in.nextInt();
            int b = in.nextInt();
            long w = (long)in.nextInt();
            
            adj[a].add(new Edge(b, w));
            adj[b].add(new Edge(a, w));

        }

        dijkstra(N, S, T);
    }
}

File: 022-SimpleDijkstras/handbook_solution.cpp
#include <bits/stdc++.h>
using std::vector;
using std::priority_queue;
using std::cin;
using std::cout;
using std::endl;
using std::set;

struct MyEntry {
    int index;
    long distance;
    MyEntry(long d, int i) {
        index = i;
        distance = d;
    }
    bool operator==(MyEntry &other) {
        return distance == other.distance && index == other.index;
    }
    bool operator>(MyEntry &other) {
        if (distance != other.distance) {
            return distance > other.distance;
        }
        return index > other.index;
    }
    bool operator<(MyEntry &other) {
        if (distance != other.distance) {
            return distance < other.distance;
        }
        return index < other.index;
    }
};

struct MyEdge {
    int index;
    long length;
    MyEdge(int i, long l) {
        index = i;
        length = l;
    }
};

int N, E, S, T;
const long INF = LONG_MAX;
vector<vector<MyEdge>> adj;

void dijkstra() {
    
    long* dist = new long[N];
    
    for (int n = 0; n < N; n++) {
        dist[n] = INF;
    }
    
    dist[S] = 0;
    
    int* prev = new int[N];

    set<MyEntry> frontier = set<MyEntry>();
    frontier.insert(MyEntry(0, S));

    while (!frontier.empty()) {

        MyEntry e = *frontier.begin();
        frontier.erase(e);

        int u = e.index;

        if (u == T) {
            break;
        }

        if (dist[u] < e.distance) {
            continue;
        }

        for (int e = 0; e < adj[u].size(); e++) {
            
            MyEdge edge = adj[u][e];
            
            int v = edge.index;
            long delta = edge.length;

            if (dist[u] + delta < dist[v]) {
                prev[v] = u;
                dist[v] = dist[u] + delta;
                frontier.insert(MyEntry(dist[v], v));
            }
        }
    }

    cout << dist[T] << endl;

    delete [] dist;
    delete [] prev;

}

int main() {
    
    cin >> N >> E >> S >> T;
    
    for (int n = 0; n < N; n++) {
        adj.push_back(vector<MyEdge>());
    }
    
    for (int e = 0; e < E; e++) {
        int a, b, w;
        cin >> a >> b >> w;
        adj[a].push_back(MyEdge(b, w));
        adj[b].push_back(MyEdge(a, w));
    }

    dijkstra();

}
